not-script/not-script.c

```diff
@@ -0,0 +1,533 @@
+                break;
+            case 'w':
+                writable = true;
+                break;
+            default:
+                len = 0;
+                break;
+            }
+            if (len != 1)
+                errx(1, "Bad data in XenStore key %s: expected 'r' or 'w'", xenstore_path_buffer);
+            free(rw);
+        }
+    }
+
+    int fd;
+    dev_t dev;
+    uint64_t diskseq;
+
+    if ((fd = open(data, (writable ? O_RDWR : O_RDONLY) | O_NOCTTY | O_CLOEXEC | O_NONBLOCK)) < 0)
+        err(1, "open(%s)", data);
+    char phys_dev[16 + 8 + 8 + 2 + 1];
+
+    process_blk_dev(fd, data, writable, &dev, &diskseq, autoclear);
+    unsigned const int l =
+        (unsigned)(autoclear ?
+                   snprintf(phys_dev, sizeof phys_dev, "%" PRIx64 "@%lx:%lx",
+                            diskseq, (unsigned long)major(dev), (unsigned long)minor(dev)) :
+                   snprintf(phys_dev, sizeof phys_dev, "%lx:%lx",
+                            (unsigned long)major(dev), (unsigned long)minor(dev)));
+
+
+    if (l >= sizeof(phys_dev))
+        err(1, "snprintf");
+
+    if (autoclear) {
+        strcpy(extra_path, opened_key);
+        if (!xs_watch(h, xenstore_path_buffer, opened_key))
+            err(1, "Cannot setup XenStore watch on %s", xenstore_path_buffer);
+    }
+
+    for (;;) {
+        xs_transaction_t t = xs_transaction_start(h);
+        if (!t)
+            err(1, "Cannot start XenStore transaction");
+
+        strcpy(extra_path, "physical-device");
+        if (!xs_write(h, t, xenstore_path_buffer, phys_dev, l))
+            err(1, "xs_write(\"%s\", \"%s\")", xenstore_path_buffer, phys_dev);
+
+        strcpy(extra_path, "physical-device-path");
+        if (!xs_write(h, t, xenstore_path_buffer, data, path_len))
+            err(1, "xs_write(\"%s\", \"%s\")", xenstore_path_buffer, data);
+
+        if (!autoclear) {
+            char hex_diskseq[17];
+            if (snprintf(hex_diskseq, sizeof(hex_diskseq), "%016llx", (unsigned long long)diskseq) != 16)
+                err(1, "snprintf");
+            strcpy(extra_path, "diskseq");
+            if (!xs_write(h, t, xenstore_path_buffer, hex_diskseq, 16))
+                err(1, "xs_write(\"%s\", \"%s\")", xenstore_path_buffer, hex_diskseq);
+        }
+
+        if (xs_transaction_end(h, t, false))
+            break;
+
+        if (errno != EAGAIN)
+            err(1, "xs_transaction_end");
+    }
+
+    if (autoclear) {
+        strcpy(extra_path, opened_key);
+        for (;;) {
+            unsigned int num, state_len;
+            char *value, **watch_res = xs_read_watch(h, &num);
+            if (!watch_res)
+                err(1, "xs_read_watch");
+            warnx("Xenstore watch for %s triggered", watch_res[0]);
+            free(watch_res);
+            value = xs_read(h, 0, xenstore_path_buffer, &state_len);
+            if (value) {
+                if (state_len != 1 || (value[0] != '1' && value[0] != '0'))
```

Comment:
This change doesn't belong to "Use different physical-device format for autoclear-enabled kernels" commit.

Reply:
But also, commits history will need a cleanup before merging anyway.